### PR TITLE
RN-345 Optimizing channel switching and general improvements

### PR DESCRIPTION
**Pull request by hmhealey**
- Changed the ChannelPostList to display its existing contents switching channels. The refresh indicator now indicates loading instead of hiding the channel contents. This drastically improves the feel of switching channels since we're no longer waiting 500ms for that to disappear
- Re-ordered the actions that take place when switching a channel. Before, it went  
        1. Mark channels read and viewed on server
        2. Close drawer
        3. After closing drawer, switch channels
    Now, it goes:
        1. Switch channels
        2. Close drawer
        3. After closing drawer, mark channels read
    Like the loader changes, this will make the UI more responsive, and it will only delay the more expensive and less important network operations
- Moved around some props and containers so that we're doing less re-rendering of higher levels components. For example, the channel screen no longer re-renders when the post draft changes.
- Made some components (notably FormattedText) extend PureComponent when they didn't before

I'm still not sold on using the refresh indicator to show that the channel is loading since it causes channels to appear to bounce up and down slightly when switching channels (because the indicator is starting to appear and then disappear immediately), even on fast connections. I was hoping to be able to delay its appearance by ~100ms or so to mitigate that, but we can't control that since it's part of the FlatList component. I have some work to do for switching teams, so I might change it then.

#### Ticket Link
https://mattermost.atlassian.net/browse/RN-345

#### Checklist
- Has UI changes

#### Device Information
This PR was tested on: iOS Simulator

### CODE REVIEW

File: app/components/channel_drawer/channel_drawer.js [hunk 1]
```diff
@@ -91,37 +85,26 @@ export default class ChannelDrawer extends PureComponent {
...
-
         if (this.state.openDrawer) {
-            this.setState({openDrawer: false});
-        }
-    };
-
-    handleDrawerCloseStart = () => {
-        if (!this.closeLeftHandle) {
-            this.closeLeftHandle = InteractionManager.createInteractionHandle();
+            // The state doesn't get updated if you swipe to close
```
Comment (enahum) [inline app/components/channel_drawer/channel_drawer.js]: what does this mean? how that is affecting the behavior? 

I think that is why we used the interaction manager

File: app/components/channel_drawer/channel_drawer.js [hunk 2]
```diff
@@ -91,37 +85,26 @@ export default class ChannelDrawer extends PureComponent {
...
-            InteractionManager.clearInteractionHandle(this.openLeftHandle);
-            this.openLeftHandle = null;
-        }
     };
 
     handleDrawerOpenStart = () => {
-        if (!this.openLeftHandle) {
-            this.openLeftHandle = InteractionManager.createInteractionHandle();
+        if (!this.state.openDrawer) {
+            // The state doesn't get updated if you swipe to open
```
Comment (enahum) [inline app/components/channel_drawer/channel_drawer.js]: same as above here

File: app/components/channel_drawer/channel_drawer.js [hunk 3]
```diff
@@ -192,42 +180,42 @@ export default class ChannelDrawer extends PureComponent {
             viewChannel
         } = actions;
 
-        markChannelAsRead(currentChannel.id);
+        markChannelAsRead(currentChannelId);
```
Comment (enahum) [inline app/components/channel_drawer/channel_drawer.js]: why this action doesn't use the `same` order as selecting channels?

File: app/components/channel_drawer/channels_list/channels_list.js [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+
+import Badge from 'app/components/badge';
+import SearchBar from 'app/components/search_bar';
+import {preventDoubleTap} from 'app/utils/tap';
+import {changeOpacity, makeStyleSheetFromTheme} from 'app/utils/theme';
+
+import FilteredList from './filtered_list';
+import List from './list';
+
+class ChannelsList extends Component {
```
Comment (enahum) [inline app/components/channel_drawer/channels_list/channels_list.js]: Shouldn't this be a PureComponent?

File: app/components/channel_drawer/channels_list/channels_list.js [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+                        count={badgeCount}
+                        minHeight={20}
+                        minWidth={20}
+                    />
+                );
+            }
+
+            switcher = (
+                <TouchableHighlight
+                    onPress={() => preventDoubleTap(onShowTeams)}
```
Comment (enahum) [inline app/components/channel_drawer/channels_list/channels_list.js]: can we make this a class function?

File: app/components/channel_drawer/channels_list/index.js [hunk 6]
```diff
@@ -1,406 +1,25 @@
...
-            color: theme.mentionColor,
-            fontSize: 14,
-            paddingVertical: 2,
-            paddingHorizontal: 4,
-            textAlign: 'center',
-            textAlignVertical: 'center'
-        }
+        ...ownProps,
+        channels: getChannelsWithUnreadSection(state),
+        channelMembers: state.entities.channels.myMembers,
```
Comment (enahum) [inline app/components/channel_drawer/channels_list/index.js]: we can use here `getMyChannelMemberships` selector

File: app/screens/channel/channel.js [hunk 7]
```diff
@@ -167,14 +156,13 @@ class Channel extends PureComponent {
     };
 
     retryLoadChannels = () => {
-        this.loadChannels(this.props.currentTeam.id);
+        this.loadChannels(this.props.currentTeamId);
```
Comment (enahum) [inline app/screens/channel/channel.js]: I also think we can improve the function for `this.loadChannels` but I could make another PR for that one

File: app/screens/channel/channel_post_list/channel_post_list.js [hunk 8]
```diff
@@ -20,7 +20,7 @@ import PostList from 'app/components/post_list';
 import PostListRetry from 'app/components/post_list_retry';
 
 const {View: AnimatedView} = Animated;
-const {height: deviceHeight, width: deviceWidth} = Dimensions.get('window');
+const {width: deviceWidth} = Dimensions.get('window');
```
Comment (enahum) [inline app/screens/channel/channel_post_list/channel_post_list.js]: can we move this cause once we add the `landscape` support in #909 this might cause problems if we don't recalculate

File: app/screens/channel/channel_post_textbox/index.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import {
+    handlePostDraftChanged
+} from 'app/actions/views/channel';
+
+import ChannelPostTextbox from './channel_post_textbox';
+
+function mapStateToProps(state, ownProps) {
+    return {
+        drafts: state.views.channel.drafts,
```
Comment (enahum) [inline app/screens/channel/channel_post_textbox/index.js]: why do we need to pass the whole drafts object instead of the one required

File: app/components/channel_drawer/channels_list/channels_list.js (hunk 4 above)
Comment (hmhealey) [reply app/components/channel_drawer/channels_list/channels_list.js]: It was just a Component before, but I can't see any reason it isn't pure. I'll change it

**Comment by enahum** [comment]
can we do something to fix the weird bounce behavior?

BTW: sometimes I load a channel I can see how the content is being changed before the drawer closes and the drawer feels sort of stuck for a little while. (this is happening while the spinner is showing at the bottom of the posts list and once it finishes the drawer continues to close).

all and all it feels way better.

PS: need rebase

File: app/components/channel_drawer/channels_list/channels_list.js (hunk 5 above)
Comment (hmhealey) [reply app/components/channel_drawer/channels_list/channels_list.js]: I'll do that in a separate PR. I'm going to make a ticket to change this everywhere since I made a `wrapWithPreventDoubleTap` that does the same thing as this, but it can be used for class functions

**Comment by hmhealey** [comment]
For the bouncing, I'm thinking of adding our own semi-transparent loading spinner in the center of the screen. That way, we'll have more control over it (to delay it being shown), and it won't cause any bounce

File: app/components/channel_drawer/channel_drawer.js (hunk 1 above)
Comment (hmhealey) [reply app/components/channel_drawer/channel_drawer.js]: It's to handle how the Drawer doesn't update this.state.openDrawer when you swipe to open/close the drawer. The code didn't change; I just added the comment since it was confusing me at first

File: app/components/channel_drawer/channel_drawer.js (hunk 3 above)
Comment (hmhealey) [reply app/components/channel_drawer/channel_drawer.js]: This action is more complicated, and it always requires that we contact the server, so I didn't want to change too much and risk breaking anything. I'll see if I can improve it slightly

File: app/components/channel_drawer/channel_drawer.js (hunk 1 above)
Comment (hmhealey) [reply app/components/channel_drawer/channel_drawer.js]: I think the InteractionManager changes were for the runAfterInteractions calls, but I'm pretty sure the Drawer code does that itself

Comment (enahum) [reply app/components/channel_drawer/channel_drawer.js]: be aware of this https://github.com/mattermost/mattermost-mobile/blob/master/app/components/channel_drawer/channel_drawer.js#L343

Comment (hmhealey) [reply app/components/channel_drawer/channel_drawer.js]: Oh, I didn't notice that. Why aren't we using their code for this?

File: app/screens/channel/channel_post_list/channel_post_list.js (hunk 8 above)
Comment (hmhealey) [reply app/screens/channel/channel_post_list/channel_post_list.js]: What's the preferred way of doing this now?

File: app/components/channel_drawer/channel_drawer.js (hunk 1 above)
Comment (enahum) [reply app/components/channel_drawer/channel_drawer.js]: We weren't because we needed to keep track of this, but perhaps that will work just fine after your changes

File: app/screens/channel/channel_post_list/channel_post_list.js (hunk 8 above)
Comment (enahum) [reply app/screens/channel/channel_post_list/channel_post_list.js]: you could either do what is being done in the #909 PR where we listen for orientation changes and update the value in the state or you could grab the value whenever the components renders, I don't think we have a preferred way but definitely we can have it as a constant that gets set one the file loads as it will never change the value

Comment (hmhealey) [reply app/screens/channel/channel_post_list/channel_post_list.js]: I'll go for the changing it on render since that's probably the cleanest. Something else we could do in the future is include it in the view store, but that could be done later

File: app/components/channel_drawer/channel_drawer.js (hunk 1 above)
Comment (hmhealey) [reply app/components/channel_drawer/channel_drawer.js]: I've switched the Drawer to register with the InteractionManager itself

**Comment by hmhealey** [comment]
@enahum Okay, this is ready to go again other than the bouncing. If this is blocking Jarred's PR, I can fix that in a separate PR
